### ISSUE TITLE
Expressions' type was not recognized on "array" type

### ISSUE THREAD

**Issue by MarcinOrlowski** [kalessil/phpinspectionsea#303]
I got this code snippet:

```
<?php
function checkDbs()
{
    $nodes_for_each_dbs = [];
    $db_names_found_on_multiple_nodes = array_filter($nodes_for_each_dbs, function($db_nodes) {
        return (count($db_nodes) > 1);
    });

    if (array_key_exists($db_id, $db_names_found_on_multiple_nodes)) {
        foreach ($db_names_found_on_multiple_nodes[ $db_id ] as $node) {
            echo $node;
        }
    }
}
```

but line "foreach ($dbs_on_multiple_nodes[ $db_id ] as $node) {" gets underlined with note "Expression type was not recognized, please check type hints":

![shot](https://cloud.githubusercontent.com/assets/8041294/25909739/67a5b1ce-35ae-11e7-8fa4-124e5a844a2a.png)

I believe this is a bug, as type should be pretty clear here.

**Comment by kalessil** [comment]
Thank you for reporting, @MarcinOrlowski. 

There is no information about the array items type declared. For human it's clear for sure, but IDE types resolving has own limitations.

This will do the treak providing needed type informattion.
```php
/* @var array[] $nodes_... */
$nodes_... = [];
```

**Comment by MarcinOrlowski** [comment]
I formerly reported this as PHPStorm issue as I thought it is their inspection. I am not sure how this all works internally but I somehow expected that you can actually draw such information i.e. from method return types, i.e. `array_filter()` returns `array` so no doubts here.

**Comment by rentalhost** [comment]
It should be analyzed, but I guess that should be reported to PS team. As `array_key_exists()` should only works on arrays, and `if()` conditional pass, then we could consider that the inner variable is an array too. The same happen for `is_iterable()`, but I yet have reported to PS and seems to be fixed.

**Comment by MarcinOrlowski** [comment]
I formerly reported this to PHPStorm team ([link](https://youtrack.jetbrains.com/issue/WI-36273)) as I thought it comes from their inspection. They pointed out here so we got kinda endless loop now :)

So to sum up. Sure I can add the PHPDoc block but I think this should not be mandatory in the case I have with this code. Now both my reports (here and PS) are closed which I think should not be the case.

**Comment by rentalhost** [comment]
This inspection is managed by this inspection, but the variable type is from IDE. As the inspection uses this information, then it reports based on what IDE says. I don't know if we can do something over it.

**Comment by kalessil** [comment]
That's might be confusing but this behavior is correct. And this is not a bug of IDE/plugin.

Let me explain: 
- array_filter in stubs returns array type -> no information if it is an array of integers or strings and etc.
- foreach on array element reported because of no detailed information available (mixed type technically) about the element type.

Does it make any sense, or perhaps it's too many work to add type hints around your project?

**Comment by MarcinOrlowski** [comment]
Ah, I now see it :) It seems I assumed that inspection complains about `$db_names_found_on_multiple_nodes` not being an array, while (after I read you comment) I realized it in fact means `$db_id` referenced element of it. Now it's clear why this happens and yes, PHPDoc is the way. Thanks.

**Comment by rentalhost** [comment]
Maybe we can reconsider somethings... See: the signature of this function is `array_key_exists($key, array $search)`. It mean that the `$search` component MUST be an array. In this case, strictly, because if I pass a non-array PHP will report with type mismatch.

Then what we could do: if inside an `if()` we use some method that "test strictly" (or maybe via phpdoc) a variable, then inside we could consider it as possible type.

**Comment by kalessil** [comment]
That's on-going topic: IMO that's the responsibility of PS types resolver - related tickets in their bug-tracker existing for years already.

**Comment by MarcinOrlowski** [comment]
> It mean that the $search component MUST be an array

@rentalhost I do not think so. While you **can** assume that return value from `array_filter()` is type `array` you unfortunatelly **cannot** do the same with any of the arguments (unless they are produced by function you can "trust" as said `array_filter()`). So `$search` is **expected** to be `array` but you still can pass anything. In short - I think you cannot draw any conclusion about type from **arguments**, unless you can what is either assigned (i.e. `$search = [];`) or where it comes from (i.e. said `array_filter()`. Unluckily even this `function x(array $search) {}` cannot be fully trusted unless you track `$search` is nowhere reassigned.

**Comment by rentalhost** [comment]
Not really. In case of variable type defined on signature, you MUST pass the same type, else you will have a PHP error, and the function [will not be processed](https://3v4l.org/Fsb3h).

**Comment by MarcinOrlowski** [comment]
That's correct, but I meant rather something like this which can easily wreck the whole thing:

```
function x(array $foo) {
    $foo = false;

    foreach ($foo as $node) {
        ...
    }
}
```

**Comment by rentalhost** [comment]
In your example you should receives a warning about not anotated type. But on what I mean is only for if().
